libs/vkd3d/command.c

```diff
@@ -9056,15 +9075,107 @@ static void STDMETHODCALLTYPE d3d12_command_list_SetSamplePositions(d3d12_comman
+            dst_resource->desc.MipLevels,
+            d3d12_resource_desc_get_layer_count(&dst_resource->desc));
+
+    if (src_rect)
+    {
+        src_offset.x = src_rect->left;
+        src_offset.y = src_rect->top;
+        src_offset.z = 0;
+        extent.width = src_rect->right - src_rect->left;
+        extent.height = src_rect->bottom - src_rect->top;
+        extent.depth = 1;
+    }
+    else
+    {
+        memset(&src_offset, 0, sizeof(src_offset));
+        vk_extent_3d_from_d3d12_miplevel(&extent, &src_resource->desc, src_subresource.mipLevel);
+    }
+
+    dst_offset.x = (int32_t)dst_x;
+    dst_offset.y = (int32_t)dst_y;
+    dst_offset.z = 0;
+
+    if (mode == D3D12_RESOLVE_MODE_AVERAGE)
+    {
+        VkImageResolve vk_image_resolve;
+        vk_image_resolve.srcSubresource = src_subresource;
+        vk_image_resolve.dstSubresource = dst_subresource;
+        vk_image_resolve.extent = extent;
+        vk_image_resolve.srcOffset = src_offset;
+        vk_image_resolve.dstOffset = dst_offset;
+        d3d12_command_list_resolve_subresource(list, dst_resource, src_resource, &vk_image_resolve, format);
+    }
+    else if (mode == D3D12_RESOLVE_MODE_DECOMPRESS)
+    {
+        /* This is a glorified copy path. The region can overlap fully, in which case we have an in-place decompress.
+         * Do nothing here. We can copy within a subresource, in which case we enter GENERAL layout.
+         * Otherwise, this can always map to vkCmdCopyImage, except for DEPTH -> COLOR copy.
+         * In this case, just use the fallback paths as is. */
+        bool writes_full_subresource;
+        bool overlapping_subresource;
+        VkImageCopy image_copy;
+
+        overlapping_subresource = dst_resource == src_resource && dst_sub_resource_idx == src_sub_resource_idx;
+
+        /* In place DECOMPRESS. No-op. */
+        if (overlapping_subresource && memcmp(&src_offset, &dst_offset, sizeof(VkOffset3D)) == 0)
+            return;
+
+        /* Cannot discard if we're copying in-place. */
+        writes_full_subresource = !overlapping_subresource &&
+                d3d12_image_copy_writes_full_subresource(dst_resource,
+                        &extent, &dst_subresource);
+
+        d3d12_command_list_track_resource_usage(list, src_resource, true);
+        d3d12_command_list_track_resource_usage(list, dst_resource, !writes_full_subresource);
+
+        image_copy.srcSubresource = src_subresource;
+        image_copy.dstSubresource = dst_subresource;
+        image_copy.srcOffset = src_offset;
+        image_copy.dstOffset = dst_offset;
+        image_copy.extent = extent;
+
+        d3d12_command_list_copy_image(list, dst_resource, dst_resource->format,
+                src_resource, src_resource->format, &image_copy,
+                writes_full_subresource, overlapping_subresource);
+    }
+    else
+    {
+        /* TODO: MIN/MAX resolve for depth/stencil. */
+        FIXME("Unsupported resolve mode: %u.\n", mode);
```

Comment:
I feel like we should probably just pass the resolve mode to `d3d12_command_list_resolve_subresource` and handle it there, there are other situations where we'd want to use render pass resolves even with `AVERAGE` and I'd like to avoid unnecessary code duplication